opm/simulators/linalg/bda/openclKernels.hpp

```diff
@@ -220,295 +220,401 @@ namespace bda
-        idx += nodesPerColorPrefix[color];
-        target_block_row += nodesPerColorPrefix[color];
-        const unsigned int lane = idx_t % warpsize;
-        const unsigned int c = (lane / bs) % bs;
-        const unsigned int r = lane % bs;
-
-        while(target_block_row < nodesPerColorPrefix[color+1]){
-            const unsigned int first_block = Lrows[target_block_row];
-            const unsigned int last_block = Lrows[target_block_row+1];
-            unsigned int block = first_block + lane / (bs*bs);
-            double local_out = 0.0;
-            if(lane < num_active_threads){
-                if(lane < bs){
-                    local_out = y[target_block_row*bs+lane];
-                }
-                for(; block < last_block; block += num_blocks_per_warp){
-                    const double x_elem = x[Lcols[block]*bs + c];
-                    const double A_elem = Lvals[block*bs*bs + c + r*bs];
-                    local_out -= x_elem * A_elem;
-                }
-            }
-
-            // do reduction in shared mem
-            tmp[lane] = local_out;
-            barrier(CLK_LOCAL_MEM_FENCE);
-
-            for(unsigned int offset = 3; offset <= 24; offset <<= 1)
+    std::string get_ILU_apply1_string(bool full_matrix) {
+        std::string s = R"(
+            __kernel void ILU_apply1(
+                __global const double *LUvals,
+                __global const unsigned int *LUcols,
+                __global const unsigned int *LUrows,
+                __global const int *diagIndex,
+                __global const double *y,
+                __global double *x,
+                __global const unsigned int *nodesPerColorPrefix,
+                const unsigned int color,
+                const unsigned int block_size,
+                __local double *tmp)
             {
-                if (lane + offset < warpsize)
-                {
-                    tmp[lane] += tmp[lane + offset];
+                const unsigned int warpsize = 32;
+                const unsigned int bs = block_size;
+                const unsigned int idx_t = get_local_id(0);
+                const unsigned int num_active_threads = (warpsize/bs/bs)*bs*bs;
+                const unsigned int num_blocks_per_warp = warpsize/bs/bs;
+                const unsigned int NUM_THREADS = get_global_size(0);
+                const unsigned int num_warps_in_grid = NUM_THREADS / warpsize;
+                unsigned int idx = get_global_id(0);
+                unsigned int target_block_row = idx / warpsize;
+                target_block_row += nodesPerColorPrefix[color];
+                const unsigned int lane = idx_t % warpsize;
+                const unsigned int c = (lane / bs) % bs;
+                const unsigned int r = lane % bs;
+
+                while(target_block_row < nodesPerColorPrefix[color+1]){
+                    const unsigned int first_block = LUrows[target_block_row];
+                    )";
+        if (full_matrix) {
```

Comment:
Do you mean for speed or otherwise? These functions are only called once or twice, and `+=` is [quite fast](https://stackoverflow.com/questions/18892281/most-optimized-way-of-concatenation-in-strings).

Reply:
indeed